### PR TITLE
Add schema cache implementation

### PR DESCRIPTION
**Pull request by sforward**
## Summary
Currently, on every single request the schema is built by instantiating all types, queries, mutations, etc, and calling all `args`, `fields` and `type` methods. This takes some time, especially when you have lots of queries and mutations. When investigating this issue, I came across a possible solution in de [docs of webonyx/graphql-php](https://webonyx.github.io/graphql-php/schema-definition-language/#performance-considerations). This works by caching the intermediate parsed representation of the schema to the filesystem. After loading from cache only the necessary classes will be resolved.

The schema cache could be enabled via the config. By default it is disabled on local environment:

```
    'schema_cache' => [
        // Enable/disable the schema cache
        'enable' => env('GRAPHQL_SCHEMA_CACHE', 'local' !== env('APP_ENV')),

        'cache_driver' => env('GRAPHQL_SCHEMA_CACHE_DRIVER', config('cache.default')),

        'cache_prefix' => env('GRAPHQL_SCHEMA_CACHE_PREFIX', 'graphql.schema_cache'),
    ],
```

If the schema cache is enabled, it could be disabled per schema via the config:

```
    'schemas' => [
        'default' => [
            'query' => [
                // ExampleQuery::class,
            ],

            // etc ...

            // Disable schema cache for this schema
            'disable_cache' => true,
        ],
    ],
```

The implementation comes with some breaking changes, so hopefully it's in time for v9. When this PR is approved, I will update the readme and the changelog.

---

Type of change:
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update
- [ ] Misc. change (internal, infrastructure, maintenance, etc.)

Checklist:
- [x] Existing tests have been adapted and/or new tests have been added
- [ ] Add a CHANGELOG.md entry
- [ ] Update the README.md
- [x] Code style has been fixed via `composer fix-style`

### CODE REVIEW

**Comment by sforward** [comment]
Thanks for the feedback! Fair point about the configurability of this cache. I will look into it soon. Besides, I experienced a problem with custom scalars that needs to be fixed too.

**Comment by sforward** [comment]
The issues with the scalars and types have been resolved now, by the new TypeConfigDecorator class. The pagination types are not defined via the config, so to resolve them consistently I have added an abstract parent. Custom or new pagination types need to extend this `AbstractPaginationType` class, to handle them with schema cache.

I have also made the schema cache configurable (I have updated the example in my original comment). By default the schema cache is disabled on local environment. It only needs to be refreshed if the schema config changes.

**Comment by websitevirtuoso** [comment]
Looks great. do we have any news about this feature?

**Comment by websitevirtuoso** [comment]
Yes Really good feature especially when schema and gql parsing consume time. my gql queries take by 1 sec. but sql queries fine by 30-50ms

**Comment by sforward** [comment]
I've improved the mapping from name to type class, when not provided via the config. This mapping is also cached now.

I needed to access the name of the queries/types, so I introduced a `getName` method. The name was already accessible via the magic getter, but this also calls several other methods, causing other problems.

File: src/Support/SchemaCache/SchemaCache.php [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+     *
+     * @return array<string, mixed>
+     */
+    protected function enrichSchemaConfig(array $schemaConfig): array
+    {
+        $namedSchemaConfig = [
+            'types' => $this->config->get('graphql.types', []),
+        ];
+
+        foreach ($schemaConfig as $group => $types) {
```
[the remark this answers is not in the document]
Comment (sforward) [reply src/Support/SchemaCache/SchemaCache.php]: I have pushed a fix and test for this config issue! Global types were already supported, see the Upload type in the SchemaCacheTest class.

**Comment by sforward** [comment]
I have not enough time to keep this going currently. I will close this PR for now, hopefully I find some time later this year.

**Pull request closed without merging**